libass/ass_fontselect.c

```diff
@@ -166,7 +166,64 @@ static ASS_FontProviderFuncs ft_funcs = {
     .destroy_font      = destroy_font_ft,
 };
 
-static void load_fonts_from_dir(ASS_Library *library, const char *dir)
+
+#if defined(_WIN32) && !defined(__CYGWIN__)
+static void load_fonts_from_dir(ASS_Library* library, const char* dir)
+{
+    _WDIR* d = NULL;
+    UINT cur_cp = CP_UTF8;
+    wchar_t* wdir = ass_convert_to_utf16le(dir, cur_cp);
+
+    if (!wdir || !(d = _wopendir(wdir))) {
+        free(wdir);
+        cur_cp = CP_ACP;
+#if WINAPI_FAMILY_PARTITION(WINAPI_PARTITION_DESKTOP)
+        if (!AreFileApisANSI()) cur_cp = CP_OEMCP;
+#endif
+        wdir = ass_convert_to_utf16le(dir, cur_cp);
+        if (wdir) d = _wopendir(wdir);
+    }
+    if (!d) {
+        free(wdir);
+        return;
+    }
+    size_t dirlen = strlen(dir);
+    size_t namemax = 0;
+    char* namebuf = NULL;
+    while (1) {
+        struct _wdirent* entry = _wreaddir(d);
+        if (!entry)
+            break;
+        char* d_name = ass_convert_from_utf16le(entry->d_name, cur_cp);
+        if (d_name[0] == '.')
+            continue;
+        size_t namelen = dirlen + strlen(d_name) + 2u;
+        if (namelen < 2 || namelen - 2 < dirlen)
+            goto clean_d_name;
+        if (namelen > namemax) {
+            size_t newlen = FFMAX(2048, namelen + FFMIN(256, SIZE_MAX - namelen));
+            if (ASS_REALLOC_ARRAY(namebuf, newlen))
+                namemax = newlen;
+            else
+                goto clean_d_name;
+        }
+        snprintf(namebuf, namemax, "%s/%s", dir, d_name);
+        size_t bufsize = 0;
+        ass_msg(library, MSGL_INFO, "Loading font file '%s'", namebuf);
+        void* data = read_file(library, namebuf, &bufsize);
```

Comment:
At this point we know that `namebuf` is UTF-8 file name, but `read_file()` also tries to open it as ANSI. That can potentially lead to undesirable side effects so it's better to avoid it. Also there's unnecessary UTF16→UTF8→UTF16 round-trip here, maybe it's worth a try to eliminate it.

Reply:
this also bothered me, but most time it won't fallback to ascii.
also, converting between utf16 and utf8 is lossless, though little inefficient.

Reply:
> Ideally there'd be no roundtrip conversion, yes, but then we'd need another `read_file` for Microsoft's `wchar_t`.

We can move `fopen()` outside of `read_file()` and make something like `read_opened_file()`.

> we pass it as UTF-8 and UTF-8 is the first one to be tested.

But opening can fail even if filename decoded correctly (access problems). And this can trigger incorrect second choice. I think there should be at least some flag to skip that second opening attempt.

> As long as we don't support WTF-8 anyway,

Implementing WTF-8 support isn't that hard so I don't see the benefit of postponing it.

Reply:
Right, opening can also fail if the file exists (or it could have been deleted in the meantime). So I'm revising my previous statement: It should use either use `wchar_t*` directly or add WTF-8-support. (If `read_file` is split though, I think this should by a separate *preceding* commit and also add an `ass_` prefix to the new functions.)

> Implementing WTF-8 support isn't that hard so I don't see the benefit of postponing it.

If someone is volunteering to do that, sure.

Reply:
About the problem that well-decoded filename failed to be opened because of some access problems, would "checking the file exsiting before opening it“ be a solution for the problem?
codes like this
```cpp
bool ass_file_exsits(const wchar_t *path) {
    DWORD file_attributes = GetFileAttributesW(path);
    if (file_attributes == INVALID_FILE_ATTRIBUTES)
    {
        switch (GetLastError())
        {
        case ERROR_FILE_NOT_FOUND:
        case ERROR_PATH_NOT_FOUND:
            return false;
        default:
            return true;
        }
    }
    return true;
}
```